driver/src/price_estimation/average_price_source.rs

```diff
@@ -0,0 +1,86 @@
+use super::{PriceSource, Token};
+use crate::models::TokenId;
+use anyhow::{anyhow, Result};
+use crossbeam_utils::thread;
+use std::collections::HashMap;
+use std::sync::Mutex;
+
+/// Combines two prices into one average.
+pub struct AveragePriceSource<T0, T1> {
+    source_0: Mutex<T0>,
+    source_1: Mutex<T1>,
+}
+
+impl<T0, T1> AveragePriceSource<T0, T1> {
+    pub fn new(source_0: T0, source_1: T1) -> Self {
+        Self {
+            source_0: Mutex::new(source_0),
+            source_1: Mutex::new(source_1),
+        }
+    }
+}
+
+impl<T0: PriceSource + Send, T1: PriceSource + Send> PriceSource for AveragePriceSource<T0, T1> {
+    fn get_prices(&self, tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
+        let prices = thread::scope(|s| {
+            let handle_0 = s.spawn(|_| self.source_0.lock().unwrap().get_prices(tokens));
+            let handle_1 = s.spawn(|_| self.source_1.lock().unwrap().get_prices(tokens));
+            (handle_0.join().unwrap(), handle_1.join().unwrap())
+        })
+        .unwrap();
```

Comment:
There is a lot of unwrap going on here. Can we translate those into errors instead of panicking? 

Also if I understand [crossbeam::thread::scope](https://docs.rs/crossbeam-utils/0.6.6/crossbeam_utils/thread/fn.scope.html) correctly, it joins all spawned thread automatically and will return an error if one of them fails (thus we shouldn't have to join and at least not unwrap the join() call manually)

Reply:
These unwraps are all related to the threads panicking. Usually you cannot do anything about panics in Rust except when they happen in another thread. So if we did the same thing we are doing in the threads which querying the prices in the main thread we would crash in a panic. Thus it makes sense to unwrap the results since that just leads to a panic again.

Reply:
The manual join is needed to get the queried prices back out of the `scope` function.

Reply:
Hope this argument makes sense. If not we can revisit it later. I'll merge now it since we wanted to have this for the weekend and I want to see if there are any problems on staging before the evening.

Reply:
Yes, makes sense